fix(upgrade): avoid memory leak when removing downgraded components

gkalpak opened the pull request:
_This is an alternative to #39921._

##
Previously, due to the way the AngularJS and Angular clean-up processes interfere with each other when removing an AngularJS element that contains a downgraded Angular component, the data associated with the host element of the downgraded component was not removed. This data was kept in an internal AngularJS cache, which prevented the element and component instance from being garbage-collected, leading to memory leaks.

This commit fixes this by ensuring the element data is explicitly removed when cleaning up a downgraded component.

NOTE:
This is essentially the equivalent of #26209 but for downgraded (instead of upgraded) components.

Fixes #39911
Closes #39921

mary-poppins commented:
You can preview 830b3b0 at https://pr39965-830b3b0.ngbuilds.io/.

mary-poppins commented:
You can preview 61baba2 at https://pr39965-61baba2.ngbuilds.io/.

petebacondarwin commented on packages/upgrade/src/common/src/downgrade_component_adapter.ts:
```diff
@@ -216,11 +216,38 @@ export class DowngradeComponentAdapter {
+        // `cleanData()` also will invoke the AngularJS `$destroy` event on the element:
+        //   https://github.com/angular/angular.js/blob/2e72ea13fa98bebf6ed4b5e3c45eaf5f990ed16f/src/Angular.js#L1932-L1945
+        angularElement.cleanData(this.element);
+        angularElement.cleanData(this.element.children!());
```
Is this enough to clean up all descendants? I.e. the ones that are children of the children?

gkalpak replied:
Yes, it is enough. I'll update the test as suggested in your other comment.

gkalpak replied:
Well, what do you know! It wsn't enough (as proved by the failing test). I fixed the code to use `querySelectorall('*')`, which is [the same as in AngularJS][1].

Great catch ❤️ 

[1]: https://github.com/angular/angular.js/blob/2e72ea13fa98bebf6ed4b5e3c45eaf5f990ed16f/src/jqLite.js#L353

petebacondarwin commented on packages/upgrade/src/dynamic/test/upgrade_spec.ts:
```diff
@@ -695,8 +695,32 @@ withEachNg1Version(() => {
              onDestroyed.subscribe(() => {
-               ref.dispose();
+               setTimeout(() => {
+                 expect(element.textContent).not.toContain('test');
+                 expect(ng2Element.data!('foo')).toBeUndefined();
+                 expect(ng2Children.data!('baz')).toBeUndefined();
+                 expect(ng2ElementDestroyed).toBe(true);
+                 expect(ng2ChildrenDestroyed).toBe(true);
```
I think it is dangerous to put these expectations inside a callback, since if the callback is never called, then we won't get a failure.
Can we move these to a deterministic place after the `destroyIt` has been set to true?

petebacondarwin commented on packages/upgrade/src/dynamic/test/upgrade_spec.ts:
```diff
@@ -678,7 +678,7 @@ withEachNg1Version(() => {
            });
 
-           @Component({selector: 'ng2', template: 'test'})
+           @Component({selector: 'ng2', template: '<span>test</span>'})
```
Should we have a grandchild element with data on it, to prove that this is also cleaned up?

gkalpak committed:
fix(upgrade): avoid memory leak when removing downgraded components

Previously, due to the way the AngularJS and Angular clean-up processes
interfere with each other when removing an AngularJS element that
contains a downgraded Angular component, the data associated with the
host element of the downgraded component was not removed. This data was
kept in an internal AngularJS cache, which prevented the element and
component instance from being garbage-collected, leading to memory
leaks.

This commit fixes this by ensuring the element data is explicitly
removed when cleaning up a downgraded component.

NOTE:
This is essentially the equivalent of #26209 but for downgraded (instead
of upgraded) components.

Fixes #39911
Closes #39921

gkalpak committed:
refactor(upgrade): remove unused parameters/properties/variables

This commit removes some unused parameters, properties and variables in
various `@angular/upgrade` functions.

gkalpak committed:
fixup! fix(upgrade): avoid memory leak when removing downgraded components

gkalpak committed:
fixup! fix(upgrade): avoid memory leak when removing downgraded components

mary-poppins commented:
You can preview 7d18ca3 at https://pr39965-7d18ca3.ngbuilds.io/.

petebacondarwin reviewed (approved):
GREAT STUFF!

mhevery commented:
[presubmit](http://test/OCL:346338966:BASE:346338984:1607447447677:570ddcd3)

mhevery commented:
[presubmit](https://tap.corp.google.com/ui#id=OCL:346338966:BASE:346338984:1607447447729:b1309055)
